compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,335 @@ fn check_for_underconstrained_values_within_function(
+    ) {
+        trace!("processing instructions of block {} of function {}", block, function);
+        let mut arguments = Vec::new();
+        let mut results = Vec::new();
+
+        for instruction in function.dfg[block].instructions() {
+            // Optimization: there is no reason to do anything until a Brillig
+            // call is encountered
+            if self.tainted.is_empty() {
+                if let Instruction::Call { func: func_id, .. } = &function.dfg[*instruction] {
+                    if let Value::Function(callee) = &function.dfg[*func_id] {
+                        if !all_functions[&callee].runtime().is_brillig() {
+                            continue;
+                        }
+                    }
+                }
+            }
+
+            arguments.clear();
+            results.clear();
+
+            // Collect non-constant instruction arguments
+            function.dfg[*instruction].for_each_value(|value_id| {
+                if function.dfg.get_numeric_constant(value_id).is_none() {
+                    arguments.push(function.dfg.resolve(value_id));
+                }
+            });
+
+            // Collect non-constant instruction results
+            for value_id in function.dfg.instruction_results(*instruction).iter() {
+                if function.dfg.get_numeric_constant(*value_id).is_none() {
+                    results.push(function.dfg.resolve(*value_id));
+                }
+            }
+
+            // Process instructions
+            match &function.dfg[*instruction] {
+                // For memory operations, we have to link up the stored value as a parent
+                // of one loaded from the same memory slot
+                Instruction::Store { address, value } => {
+                    self.memory_slots.insert(*address, function.dfg.resolve(*value));
+                }
+                Instruction::Load { address } => {
+                    // Recall the value stored at address as parent for the results
+                    if let Some(value_id) = self.memory_slots.get(address) {
+                        self.update_children(&[function.dfg.resolve(*value_id)], &results);
+                    } else {
+                        debug!("load instruction {} has attempted to access previously unused memory location, skipping",
+                            instruction);
+                    }
+                }
+                // Check the constrain instruction arguments against those
+                // involved in Brillig calls, remove covered calls
+                Instruction::Constrain(value_id1, value_id2, _) => {
+                    self.clear_constrained(&[
+                        function.dfg.resolve(*value_id1),
+                        function.dfg.resolve(*value_id2),
+                    ]);
+                }
+                // Consider range check to also be constraining
+                Instruction::RangeCheck { value, .. } => {
```

Comment:
What if a range check were "empty," i.e. it checked that a value is within `0..Field::MAX`?

Reply:
These are automatically ignored when being added: https://github.com/noir-lang/noir/blob/dc96de7c7a3d3c325b955edb277534653204cc5a/compiler/noirc_evaluator/src/ssa/ir/instruction.rs#L855

Reply:
Should be okay then